### PR TITLE
Add result and condition filters to test results list

### LINKED ISSUE #6303: Condition and result filters on results table
### Description

Add condition and result filters for the new results table. 

Figma - https://www.figma.com/file/LglkDsGZDKMZm3XCkJPhfR/SimpleReport-2023%2B?type=design&node-id=1380-32170&mode=design&t=zqrGkZ0rm0PjuqCe-0

### Acceptance criteria
- Condition filter filters condition column (whatever disease types we have in the backend in disease table - so we don't have to change when we add new diseases)
- Test result filter filters result column (pos, neg, inc)

### Notes
Use this query to grab supported diseases - https://github.com/CDCgov/prime-simplereport/blob/9e2b1ae91c932e8d7829ee3b03f918b2e4f894d6/backend/src/main/resources/graphql/main.graphqls#L372

### PR DESCRIPTION
**Pull request by nathancrtr**
## Related Issue
- Closes #6303 

## Changes Proposed
- Adds `condition` and `result` filters for test results list
- To enable filtering on these values, the test results list now queries the GraphQL endpoint for _results_ rather than _test events_

## Additional Information
- This is a follow-up to #6104
    - This work unrolled individual multiplex results for a test event into separate rows on the result table
    - Filtering worked fine for values that do not change between results within a test (patient name, test date, etc)
    - However, since `result` and `condition` vary within a single test event, these filters had to be updated for more granular access
- Adds a new resolver and associated GraphQL schema and query definitions
- Adds new service code and `Specification`s to apply filters from user to database retrieval

#### Name filter

https://github.com/CDCgov/prime-simplereport/assets/27730981/76bdbbb0-897d-4d62-9956-535bc8c01ff9

### Date filters

https://github.com/CDCgov/prime-simplereport/assets/27730981/6343bd53-e621-4923-b136-002931b7fc0c

### Condition filter

https://github.com/CDCgov/prime-simplereport/assets/27730981/6bd8d993-4c87-4cd7-8fad-4086e302331a

### Result filter

https://github.com/CDCgov/prime-simplereport/assets/27730981/8b8ad220-85cd-4b67-b34c-ccc53ae6755d

### Role filter

https://github.com/CDCgov/prime-simplereport/assets/27730981/970c42c4-7bc0-4dfd-bf12-aebdf89b79d5

### Facility filter

https://github.com/CDCgov/prime-simplereport/assets/27730981/efa57eb1-a5f4-451c-9f6a-b07f56215f5e

### Multiple filters applied; clearing filters

https://github.com/CDCgov/prime-simplereport/assets/27730981/d5a23017-c28d-4d5f-9e32-9746023c890b

### CODE REVIEW

File: backend/src/main/resources/graphql/main.graphqls [hunk 1]
```diff
@@ -262,6 +264,8 @@ type TestOrder {
   dateTested: DateTime
   correctionStatus: String
   reasonForCorrection: String
+  createdBy: ApiUser
+  patientLink: PatientLink
```
Comment (zdeveloper) [inline backend/src/main/resources/graphql/main.graphqls:268]: whats the purpose of adding extra fields to the TestOrder?

File: backend/src/main/java/gov/cdc/usds/simplereport/service/ResultService.java [hunk 2]
```diff
@@ -16,6 +35,140 @@
...
+                            testEventJoin.get(AuditedEntity_.createdAt), endDate))));
+      }
+
+      return p;
+    };
+  }
+
+  @Transactional(readOnly = true)
+  @AuthorizationConfiguration.RequirePermissionViewAllFacilityResults
+  public Page<Result> getOrganizationResults(
```
Comment (zdeveloper) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/ResultService.java:132]: remove `getOrganizationResults` and use `getFacilityResults` since it can handle when facility is null

File: backend/src/main/java/gov/cdc/usds/simplereport/api/testresult/ResultResolver.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    if (pageSize < 1) {
+      pageSize = TestOrderService.DEFAULT_PAGINATION_PAGESIZE;
+    }
+
+    SupportedDisease supportedDisease =
+        disease != null ? diseaseService.getDiseaseByName(disease) : null;
+
+    if (facilityId == null) {
```
Comment (zdeveloper) [inline backend/src/main/java/gov/cdc/usds/simplereport/api/testresult/ResultResolver.java:47]: remove if statement and only use `getFacilityResults`

File: frontend/src/app/commonComponents/MultiplexResultsGuidance.tsx [hunk 4]
```diff
@@ -5,7 +5,7 @@ import CovidResultGuidance from "../commonComponents/CovidResultGuidance";
 import {
   hasPositiveFluResults,
   getResultByDiseaseName,
-  haCovidResults,
+  hasCovidResults,
```
Comment (zdeveloper) [inline frontend/src/app/commonComponents/MultiplexResultsGuidance.tsx:8]: hahahhahaCovidResults :D

File: frontend/src/app/testResults/TestResultsList.test.tsx [hunk 5]
```diff
@@ -279,6 +279,28 @@ describe("TestResultsList", () => {
...
+      const { user } = renderWithUser();
+      expect(
+        await screen.findByText("Test Results", { exact: false })
+      ).toBeInTheDocument();
+      expect(
+        await screen.findByText("Cragell, Barb Whitaker")
+      ).toBeInTheDocument();
+      expect(await screen.findByText("Gerard, Sam G"));
+      expect(
+        await screen.findByRole("option", { name: "Negative" })
```
Comment (zdeveloper) [inline frontend/src/app/testResults/TestResultsList.test.tsx:293]: remove `expect(await screen.findByRole("option", { name: "Negative" })`
since it doesnt add value to the test

File: frontend/src/app/testResults/TestResultsList.tsx [hunk 6]
```diff
@@ -520,6 +525,27 @@ export const DetachedTestResultsList = ({
...
+                    label: MULTIPLEX_DISEASES.FLU_A,
+                  },
+                  {
+                    value: MULTIPLEX_DISEASES.FLU_B,
+                    label: MULTIPLEX_DISEASES.FLU_B,
+                  },
+                ]}
+                defaultSelect
+                onChange={setFilterParams("disease")}
+              />
```
Comment (zdeveloper) [inline frontend/src/app/testResults/TestResultsList.tsx:548]: Per story AC, the values here should be loaded from the backend so we dont have to keep adding more values as we add more diseases, doesnt have to block this PR, but it could be added as a followup pr as part of the same ticket

File: frontend/src/app/testResults/operations.graphql [hunk 7]
```diff
@@ -98,39 +98,49 @@ query GetFacilityResultsForCsvWithCount(
...
+      }
       deviceType {
         internalId
         name
+#        manufacturer
+#        model
+#        swabTypes {
+#          internalId
+#          name
+#        }
```
Comment (zdeveloper) [inline frontend/src/app/testResults/operations.graphql:143]: remove commented out code

File: frontend/src/app/testResults/resultsTable/ResultsTable.test.tsx [hunk 8]
```diff
@@ -123,24 +124,18 @@ describe("Component ResultsTable", () => {
...
+
+      expect(screen.getByTestId(testId)).toBeInTheDocument();
     }
 
-    // TODO: all expected elements are present, but is it exhaustive?
-    // Test to check the # of rows
     expect(screen.getByTestId("filtered-results").children.length).toBe(9);
   });
 
+  /*
```
Comment (zdeveloper) [inline frontend/src/app/testResults/resultsTable/ResultsTable.test.tsx:138]: this test has been commented out, please uncomment or remove if it doesnt apply anymore

File: frontend/src/app/testResults/resultsTable/ResultsTable.tsx [hunk 9]
```diff
@@ -169,94 +130,97 @@ const generateResultRows = (
...
+        )}
+        data-testid={`test-result-${r.id}-${diseaseIdName}`}
+        data-patient-link={null}
+        /*
+        data-patient-link={
+          r.testOrder.patientLink
+            ? `${getUrl()}pxp?plid=${r.testOrder.patientLink.internalId}`
+            : null
+        }
+         */
```
Comment (zdeveloper) [inline frontend/src/app/testResults/resultsTable/ResultsTable.tsx:174]: remove this commented out code

**Review by zdeveloper — COMMENTED** [review]
LGTU, paired on the review with @mpbrown, some minor suggestions and we believe we need a follow up pr to no hard code the diseases on the disease filter list

File: backend/src/main/resources/graphql/main.graphqls [hunk 10]
```diff
@@ -288,9 +292,33 @@ type TestResultsPage {
...
+type Result {
+  id: ID!
+  facility: Facility!
+  patient: Patient!
+  dateAdded: String!
+  dateUpdated: DateTime
+  pregnancy: String
+  noSymptoms: Boolean!
+  symptoms: String!
+  symptomOnset: LocalDate
```
Comment (zdeveloper) [inline backend/src/main/resources/graphql/main.graphqls:304]: do we need AOE answers on the results object? (pregnancy, noSymptoms, symptoms, symptomOnset)

File: backend/src/main/java/gov/cdc/usds/simplereport/service/ResultService.java (hunk 2 above)
Comment (nathancrtr) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/ResultService.java:132]: The two methods are configured with different authorizations requirements:

| Method | Auth |
|--------|--------|
| getFacilityResults | RequirePermissionReadResultListAtFacility |
| getOrganizationResults | RequirePermissionViewAllFacilityResults | 

Is there a clean way of conditionally applying those requirements within a single method, or shall I keep it as-is?

File: frontend/src/app/testResults/TestResultsList.tsx (hunk 6 above)
Comment (nathancrtr) [reply frontend/src/app/testResults/TestResultsList.tsx:548]: I'm honestly not sure if I'm sold on pulling those values from the backend here to begin with. Adding new diseases to SimpleReport will already involve touchpoints in the frontend code, like [our constants](https://github.com/CDCgov/prime-simplereport/blob/fbdd80caa304a9d959b22c4f5a3d08731eb17877/frontend/src/app/testResults/constants.ts#L1-L5).

Also, `HIV` and `RSV` have already been added to the database but we're not ready to expose those values to the disease filter, so we'd need to turn around and hardcode _those_ values somewhere instead.

Basically I'm leaning towards there being a logical distinction between `diseases that SR "knows about"` and `diseases that users can submit for`, at least at this time.

Whatchy'all think?

File: backend/src/main/resources/graphql/main.graphqls (hunk 10 above)
Comment (nathancrtr) [reply backend/src/main/resources/graphql/main.graphqls]: Nah, removed in https://github.com/CDCgov/prime-simplereport/pull/6682/commits/34346c5a58c35cca7a4fde47695fcf9e994ce8ae.

File: frontend/src/app/testResults/resultsTable/ResultsTable.tsx (hunk 9 above)
Comment (nathancrtr) [reply frontend/src/app/testResults/resultsTable/ResultsTable.tsx]: 👍 Removed in https://github.com/CDCgov/prime-simplereport/pull/6682/commits/34346c5a58c35cca7a4fde47695fcf9e994ce8ae.

File: frontend/src/app/testResults/resultsTable/ResultsTable.test.tsx (hunk 8 above)
Comment (nathancrtr) [reply frontend/src/app/testResults/resultsTable/ResultsTable.test.tsx]: 👍 removed in https://github.com/CDCgov/prime-simplereport/pull/6682/commits/34346c5a58c35cca7a4fde47695fcf9e994ce8ae.

File: frontend/src/app/testResults/operations.graphql (hunk 7 above)
Comment (nathancrtr) [reply frontend/src/app/testResults/operations.graphql]: 👍 removed in https://github.com/CDCgov/prime-simplereport/pull/6682/commits/34346c5a58c35cca7a4fde47695fcf9e994ce8ae.

File: frontend/src/app/testResults/TestResultsList.test.tsx (hunk 5 above)
Comment (nathancrtr) [reply frontend/src/app/testResults/TestResultsList.test.tsx]: Good catch 👁️ https://github.com/CDCgov/prime-simplereport/pull/6682/commits/cc7f1bf0b930c9d484e035cd2a54f991141b303f

File: backend/src/main/resources/graphql/main.graphqls (hunk 1 above)
Comment (nathancrtr) [reply backend/src/main/resources/graphql/main.graphqls]: Ain't none, removed in https://github.com/CDCgov/prime-simplereport/pull/6682/commits/34346c5a58c35cca7a4fde47695fcf9e994ce8ae.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/ResultService.java (hunk 2 above)
Comment (mpbrown) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/ResultService.java:132]: I think it's good as is if they have different auth requirements

File: frontend/src/app/testResults/TestResultsList.tsx (hunk 6 above)
Comment (mpbrown) [reply frontend/src/app/testResults/TestResultsList.tsx:548]: Hmm that's a good point where the frontend already requires some hardcoded reference for the diseases. At our next pairing session I'm curious to get your thoughts on this as it relates to what I'm working on with adding HIV support since it seems like we'll always need some way to distinguish which AOE questions to show

**Review by mpbrown — APPROVED** [review]
LGTM! Nice work on this!

**Review by DanielSass — DISMISSED** [review]
lgtm

**Review by DanielSass — APPROVED** [review]

**Review by mpbrown — APPROVED** [review]

**Comment by emyl3** [comment]
@nathancrtr I noticed something interesting with this new functionality... if we correct a result is it now expected that we see the previously corrected test in the results page along with the newly submitted result?

To repro:
- submit a test
- correct result -> click something like "Incorrect test result"
- see that it is added to the queue
- submit the corrected result
- go to the results page and see both the previous test and newly corrected test in the results
